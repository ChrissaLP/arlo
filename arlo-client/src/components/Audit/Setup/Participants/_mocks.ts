const jurisdictionFile = new File(
  [
    'Jurisdiction","Admin Email"',
    '"Death Star","wtarkin@empire.gov"',
    '"Hoth","admin@rebelalliance.ninja"',
    '"Tatooine","jabba@hutt.biz"',
  ],
  'jurisdictions.csv',
  { type: 'text/csv' }
)

export default jurisdictionFile
