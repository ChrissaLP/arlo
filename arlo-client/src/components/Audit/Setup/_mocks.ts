import { ElementType } from '../../../types'
import { setupStages } from './index'
import { ISidebarMenuItem } from '../../Atoms/Sidebar'

const relativeStages = (
  stage: ElementType<typeof setupStages>,
  state: ISidebarMenuItem['state'] = 'live'
) => {
  const prevTitle = setupStages[setupStages.indexOf(stage) - 1]
  const prevStage: ISidebarMenuItem = {
    title: prevTitle,
    active: false,
    activate: jest.fn(),
    state,
  }
  const nextTitle = setupStages[setupStages.indexOf(stage) + 1]
  const nextStage: ISidebarMenuItem = {
    title: nextTitle,
    active: false,
    activate: jest.fn(),
    state,
  }
  return { prevStage, nextStage }
}

export default relativeStages
