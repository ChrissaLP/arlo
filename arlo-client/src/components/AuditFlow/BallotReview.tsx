import React from 'react'
import { H3, Button } from '@blueprintjs/core'
import styled from 'styled-components'
import { IReview } from '../../types'
import { BallotRow, FormBlock, ProgressActions, FlushDivider } from './Atoms'
import FormButton from '../Form/FormButton'

const Wrapper = styled.div`
  display: flex;
  justify-content: space-between;
  padding: 20px 0;
  @media (max-width: 775px) {
    flex-direction: column;
  }
`

const LockedButton = styled(FormButton)`
  text-align: center;
`

interface IProps {
  contestName: string
  goAudit: () => void
  review: IReview
  nextBallot: () => void
  previousBallot: () => void
  submitBallot: (data: IReview) => void
}

const BallotReview: React.FC<IProps> = ({
  contestName,
  goAudit,
  review: { vote, comment },
  review,
  nextBallot,
  previousBallot,
  submitBallot,
}: IProps) => {
  const completeVote = vote as Exclude<IReview['vote'], null>
  /* eslint-disable no-console */
  const handleSubmit = async () => {
    await submitBallot(review)
    goAudit()
    nextBallot()
  }
  return (
    <BallotRow>
      <div className="ballot-side"></div>
      <div className="ballot-main">
        <FormBlock>
          <H3>{contestName}</H3>
          <FlushDivider />
          <Wrapper>
            {/* <ButtonGroup fill large vertical> */}
            <LockedButton disabled large intent="primary">
              {completeVote}
            </LockedButton>
            <Button icon="edit" minimal onClick={goAudit}>
              Edit
            </Button>
            {/* </ButtonGroup> */}
          </Wrapper>
          <p>{comment && `COMMENT: ${comment}`}</p>
        </FormBlock>
        <ProgressActions>
          <FormButton type="submit" onClick={handleSubmit} intent="success">
            Submit &amp; Next Ballot
          </FormButton>
          <Button onClick={previousBallot} minimal>
            Back
          </Button>
        </ProgressActions>
      </div>
    </BallotRow>
  )
}

export default BallotReview
