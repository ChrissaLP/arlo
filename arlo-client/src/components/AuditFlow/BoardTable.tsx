import React from 'react'
import styled from 'styled-components'

const P = styled.p`
  margin-top: 100px;
`

interface Props {
  setIsLoading: (arg0: boolean) => void
  isLoading: boolean
}

const BoardTable: React.FC<Props> = ({  }: Props) => {
  return <P>Board Table</P>
}

export default BoardTable
