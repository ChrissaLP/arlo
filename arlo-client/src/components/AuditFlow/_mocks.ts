import { IAuditBoard } from '../../types'

export const dummyBoard: IAuditBoard[] = [
  {
    id: '123',
    name: 'Audit Board #1',
    members: [],
  },
  {
    id: '123',
    name: 'Audit Board #1',
    members: [
      {
        name: 'John Doe',
        affiliation: '',
      },
      {
        name: 'Jane Doe',
        affiliation: 'LIB',
      },
    ],
  },
  {
    id: '123',
    name: 'Audit Board #1',
    members: [
      {
        name: 'John Doe',
        affiliation: '',
      },
      {
        name: 'Jane Doe',
        affiliation: 'LIB',
      },
    ],
    ballots: [
      {
        tabulator: '11',
        batch: '0003-04-Precinct 13 (Jonesboro Fire Department)',
        position: '313',
        status: 'NOT_AUDITED',
        vote: null,
        comment: '',
        id: '1',
      },
      {
        tabulator: '17',
        batch: '0003-04-Precinct 19 (Jonesboro Fire Department)',
        position: '2112',
        status: 'NOT_AUDITED',
        vote: null,
        comment: '',
        id: '2',
      },
      {
        tabulator: '23',
        batch: '0003-04-Precinct 29 (Jonesboro Fire Department)',
        position: '1789',
        status: 'NOT_AUDITED',
        vote: null,
        comment: '',
        id: '3',
      },
      {
        tabulator: '17',
        batch: '0003-04-Precinct 19 (Jonesboro Fire Department)',
        position: '2112',
        status: 'NOT_AUDITED',
        vote: null,
        comment: '',
        id: '4',
      },
      {
        tabulator: '23',
        batch: '0003-04-Precinct 29 (Jonesboro Fire Department)',
        position: '1789',
        status: 'AUDITED',
        vote: null,
        comment: '',
        id: '5',
      },
    ],
  },
  {
    id: '123',
    name: 'Audit Board #1',
    members: [
      {
        name: 'John Doe',
        affiliation: '',
      },
      {
        name: 'Jane Doe',
        affiliation: 'LIB',
      },
    ],
    ballots: [
      {
        tabulator: '11',
        batch: '0003-04-Precinct 13 (Jonesboro Fire Department)',
        position: '313',
        status: 'AUDITED',
        vote: null,
        comment: '',
        id: '1',
      },
      {
        tabulator: '17',
        batch: '0003-04-Precinct 19 (Jonesboro Fire Department)',
        position: '2112',
        status: 'AUDITED',
        vote: null,
        comment: '',
        id: '2',
      },
      {
        tabulator: '23',
        batch: '0003-04-Precinct 29 (Jonesboro Fire Department)',
        position: '1789',
        status: 'AUDITED',
        vote: null,
        comment: '',
        id: '3',
      },
      {
        tabulator: '17',
        batch: '0003-04-Precinct 19 (Jonesboro Fire Department)',
        position: '2112',
        status: 'AUDITED',
        vote: null,
        comment: '',
        id: '4',
      },
      {
        tabulator: '23',
        batch: '0003-04-Precinct 29 (Jonesboro Fire Department)',
        position: '1789',
        status: 'AUDITED',
        vote: null,
        comment: '',
        id: '5',
      },
      {
        tabulator: '11',
        batch: '0003-04-Precinct 13 (Jonesboro Fire Department)',
        position: '313',
        status: 'AUDITED',
        vote: null,
        comment: '',
        id: '6',
      },
      {
        tabulator: '17',
        batch: '0003-04-Precinct 19 (Jonesboro Fire Department)',
        position: '2112',
        status: 'AUDITED',
        vote: null,
        comment: '',
        id: '7',
      },
      {
        tabulator: '23',
        batch: '0003-04-Precinct 29 (Jonesboro Fire Department)',
        position: '1789',
        status: 'AUDITED',
        vote: null,
        comment: '',
        id: '8',
      },
      {
        tabulator: '17',
        batch: '0003-04-Precinct 19 (Jonesboro Fire Department)',
        position: '2112',
        status: 'AUDITED',
        vote: null,
        comment: '',
        id: '9',
      },
      {
        tabulator: '23',
        batch: '0003-04-Precinct 29 (Jonesboro Fire Department)',
        position: '1789',
        status: 'AUDITED',
        vote: null,
        comment: '',
        id: '10',
      },
      {
        tabulator: '11',
        batch: '0003-04-Precinct 13 (Jonesboro Fire Department)',
        position: '313',
        status: 'AUDITED',
        vote: null,
        comment: '',
        id: '11',
      },
      {
        tabulator: '17',
        batch: '0003-04-Precinct 19 (Jonesboro Fire Department)',
        position: '2112',
        status: 'AUDITED',
        vote: null,
        comment: '',
        id: '12',
      },
      {
        tabulator: '23',
        batch: '0003-04-Precinct 29 (Jonesboro Fire Department)',
        position: '1789',
        status: 'AUDITED',
        vote: null,
        comment: '',
        id: '13',
      },
      {
        tabulator: '17',
        batch: '0003-04-Precinct 19 (Jonesboro Fire Department)',
        position: '2112',
        status: 'AUDITED',
        vote: null,
        comment: '',
        id: '14',
      },
      {
        tabulator: '23',
        batch: '0003-04-Precinct 29 (Jonesboro Fire Department)',
        position: '1789',
        status: 'AUDITED',
        vote: null,
        comment: '',
        id: '15',
      },
      {
        tabulator: '11',
        batch: '0003-04-Precinct 13 (Jonesboro Fire Department)',
        position: '313',
        status: 'AUDITED',
        vote: null,
        comment: '',
        id: '16',
      },
      {
        tabulator: '17',
        batch: '0003-04-Precinct 19 (Jonesboro Fire Department)',
        position: '2112',
        status: 'AUDITED',
        vote: null,
        comment: '',
        id: '17',
      },
      {
        tabulator: '23',
        batch: '0003-04-Precinct 29 (Jonesboro Fire Department)',
        position: '1789',
        status: 'AUDITED',
        vote: null,
        comment: '',
        id: '18',
      },
      {
        tabulator: '17',
        batch: '0003-04-Precinct 19 (Jonesboro Fire Department)',
        position: '2112',
        status: 'AUDITED',
        vote: null,
        comment: '',
        id: '19',
      },
      {
        tabulator: '23',
        batch: '0003-04-Precinct 29 (Jonesboro Fire Department)',
        position: '1789',
        status: 'AUDITED',
        vote: null,
        comment: '',
        id: '20',
      },
      {
        tabulator: '11',
        batch: '0003-04-Precinct 13 (Jonesboro Fire Department)',
        position: '313',
        status: 'AUDITED',
        vote: null,
        comment: '',
        id: '21',
      },
      {
        tabulator: '17',
        batch: '0003-04-Precinct 19 (Jonesboro Fire Department)',
        position: '2112',
        status: 'AUDITED',
        vote: null,
        comment: '',
        id: '22',
      },
      {
        tabulator: '23',
        batch: '0003-04-Precinct 29 (Jonesboro Fire Department)',
        position: '1789',
        status: 'AUDITED',
        vote: null,
        comment: '',
        id: '23',
      },
      {
        tabulator: '17',
        batch: '0003-04-Precinct 19 (Jonesboro Fire Department)',
        position: '2112',
        status: 'AUDITED',
        vote: null,
        comment: '',
        id: '24',
      },
      {
        tabulator: '23',
        batch: '0003-04-Precinct 29 (Jonesboro Fire Department)',
        position: '1789',
        status: 'AUDITED',
        vote: null,
        comment: '',
        id: '25',
      },
      {
        tabulator: '11',
        batch: '0003-04-Precinct 13 (Jonesboro Fire Department)',
        position: '313',
        status: 'AUDITED',
        vote: null,
        comment: '',
        id: '26',
      },
      {
        tabulator: '17',
        batch: '0003-04-Precinct 19 (Jonesboro Fire Department)',
        position: '2112',
        status: 'AUDITED',
        vote: null,
        comment: '',
        id: '27',
      },
      {
        tabulator: '23',
        batch: '0003-04-Precinct 29 (Jonesboro Fire Department)',
        position: '1789',
        status: 'AUDITED',
        vote: null,
        comment: '',
        id: '28',
      },
      {
        tabulator: '17',
        batch: '0003-04-Precinct 19 (Jonesboro Fire Department)',
        position: '2112',
        status: 'AUDITED',
        vote: null,
        comment: '',
        id: '29',
      },
      {
        tabulator: '23',
        batch: '0003-04-Precinct 29 (Jonesboro Fire Department)',
        position: '1789',
        status: 'AUDITED',
        vote: null,
        comment: '',
        id: '30',
      },
      {
        tabulator: '11',
        batch: '0003-04-Precinct 13 (Jonesboro Fire Department)',
        position: '313',
        status: 'AUDITED',
        vote: null,
        comment: '',
        id: '31',
      },
      {
        tabulator: '17',
        batch: '0003-04-Precinct 19 (Jonesboro Fire Department)',
        position: '2112',
        status: 'AUDITED',
        vote: null,
        comment: '',
        id: '32',
      },
      {
        tabulator: '23',
        batch: '0003-04-Precinct 29 (Jonesboro Fire Department)',
        position: '1789',
        status: 'AUDITED',
        vote: null,
        comment: '',
        id: '33',
      },
      {
        tabulator: '17',
        batch: '0003-04-Precinct 19 (Jonesboro Fire Department)',
        position: '2112',
        status: 'AUDITED',
        vote: null,
        comment: '',
        id: '34',
      },
      {
        tabulator: '23',
        batch: '0003-04-Precinct 29 (Jonesboro Fire Department)',
        position: '1789',
        status: 'AUDITED',
        vote: null,
        comment: '',
        id: '35',
      },
      {
        tabulator: '11',
        batch: '0003-04-Precinct 13 (Jonesboro Fire Department)',
        position: '313',
        status: 'AUDITED',
        vote: null,
        comment: '',
        id: '36',
      },
      {
        tabulator: '17',
        batch: '0003-04-Precinct 19 (Jonesboro Fire Department)',
        position: '2112',
        status: 'AUDITED',
        vote: null,
        comment: '',
        id: '37',
      },
      {
        tabulator: '23',
        batch: '0003-04-Precinct 29 (Jonesboro Fire Department)',
        position: '1789',
        status: 'AUDITED',
        vote: null,
        comment: '',
        id: '38',
      },
      {
        tabulator: '17',
        batch: '0003-04-Precinct 19 (Jonesboro Fire Department)',
        position: '2112',
        status: 'AUDITED',
        vote: null,
        comment: '',
        id: '39',
      },
      {
        tabulator: '23',
        batch: '0003-04-Precinct 29 (Jonesboro Fire Department)',
        position: '1789',
        status: 'AUDITED',
        vote: null,
        comment: '',
        id: '40',
      },
    ],
  },
]

export default {
  dummyBoard,
}
