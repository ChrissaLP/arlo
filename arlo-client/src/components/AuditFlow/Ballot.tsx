import React, { useState } from 'react'
import { H1, H3, Callout, H4, Divider } from '@blueprintjs/core'
import styled from 'styled-components'
import BallotAudit from './BallotAudit'
import BallotReview from './BallotReview'
import { AuditBoard, Ballot } from '../../types'

const Wrapper = styled.div`
  display: flex;
  flex-direction: column;
`

const MainCallout = styled(Callout)`
  width: 400px;
  font-weight: 700;
  background-color: #202B33;
  color: #F5F8FA;
`

export const BallotRow = styled.div`
  display: flex;
  justify-content: flex-start;
  margin: 20px 0;

  .ballot-side {
    width: 200px;
    padding: 20px 0;
  }
  .ballot-main {
    width: 50%;
    padding: 20px;
  }
`

interface Props {
  roundId: string
  ballotId: string
  board: AuditBoard
}

const Ballot: React.FC<Props> = ({ roundId, ballotId, board }: Props) => {
  const [auditing, setAuditing] = useState(true)
  const [vote, setVote] = useState(null as Ballot["vote"])

  if (!board.ballots) return null // TODO handle informatively
  const ballot = board.ballots[Number(ballotId) - 1]
  if (!ballot) return null // TODO handle informatively
  setVote(ballot.vote)

  return (
    <Wrapper>
      <H1>{board.name}: Ballot Card Data Entry</H1>
      <H3>Enter Ballot Information</H3>
      <MainCallout icon={null} intent="primary">
        Round {roundId}: auditing ballot {ballotId} of {board.ballots.length}
      </MainCallout>
      <BallotRow>
        <div className="ballot-side">
          <H4>Current ballot:</H4>
          <div>Tabulator: {ballot.tabulator}</div>
          <div>Batch: {ballot.batch}</div>
          <div>Record/Position: {ballot.position}</div>
        </div>
        <Divider />
        <div className="ballot-main">
          <H4>Are you looking at the correct ballot?</H4>
          <p>
            Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do
            eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim
            ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut
            aliquip ex ea commodo consequat. Duis aute irure dolor in
            reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla
            pariatur. Excepteur sint occaecat cupidatat non proident, sunt in
            culpa qui officia deserunt mollit anim id est laborum.
          </p>
        </div>
      </BallotRow>
      {auditing ? (
        <BallotAudit vote={vote} setVote={setVote} review={() => setAuditing(false)} />
      ) : (
        <BallotReview vote={vote} audit={() => setAuditing(true)} />
      )}
    </Wrapper>
  )
}

export default Ballot
