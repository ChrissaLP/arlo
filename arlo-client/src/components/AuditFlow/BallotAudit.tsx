import React from 'react'
import { Ballot } from '../../types';
import { Formik, FormikProps, Form, Field } from 'formik';
import { BallotRow } from './Ballot'
import { H4, H3, Divider } from '@blueprintjs/core';
import styled from 'styled-components';

const FormBlock = styled(Form)`
  background-color: #CED9E0;
`

const Option = styled(Field)`

`

interface Props {
  review: () => void
  vote: Ballot["vote"]
  setVote: (arg0: Ballot["vote"]) => void
}

interface Options {
  vote: Ballot["vote"]
  yes: boolean
  no: boolean
  noConsensus: boolean
  noVote: boolean
}

const BallotAudit: React.FC<Props> = ({ review, vote, setVote }: Props) => {
  return (
    <BallotRow>
      <div className="ballot-side"></div>
      <div className="ballot-main">
        <H4>Are you looking at the right ballot?</H4>
        <p>Step 1:  Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.</p>
        <p>Step 2: Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.</p>
        <Formik
          initialValues={{
            vote,
            yes: false,
            no: false,
            noConsensus: false,
            noVote: false
          }}
          onSubmit={({ vote }) => {
            setVote(vote)
            review()
          }}
          render={({
            handleChange,
            values,
            setFieldValue
          }: FormikProps<Options>) => {
            const handleOptionChange = (o: string) => (e: React.FormEvent) => {
              setFieldValue("vote", o);
              ['yes', 'no', 'noConsensus', 'noVote'].forEach(f => setFieldValue(f, false))
              setFieldValue(o, true)
            }
            return (
              <FormBlock>
                <H3>[insert name of choice here]</H3>
                <Divider />
                <Option name="yes" onChange={handleOptionChange("yes")} />
              </FormBlock>
            )
          }}
        />
      </div>
    </BallotRow>
  )
}

export default BallotAudit
