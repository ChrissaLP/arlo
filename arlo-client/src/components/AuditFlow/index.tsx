import React, { useState, useEffect, useCallback } from 'react'
import { Spinner } from '@blueprintjs/core'
import { Route, Switch } from 'react-router-dom'
import { History } from 'history'
import { IAuditFlowParams, IAudit, IAuditBoard, IBallot } from '../../types'
import { api } from '../utilities'
import { statusStates } from '../AuditForms/_mocks'
import {
  // dummyBoard,
  dummyBallots,
} from './_mocks'
import BoardTable from './BoardTable'
import MemberForm from './MemberForm'
import Ballot from './Ballot'
import Wrapper from '../Atoms/Wrapper'

interface IProps {
  match: {
    params: IAuditFlowParams
    url: string
  }
  history: History
}

const AuditFlow: React.FC<IProps> = ({
  match: {
    params: { electionId, token },
    url,
  },
  history,
}: IProps) => {
  const [isLoading, setIsLoading] = useState<boolean>(false)

  const [audit, setAudit] = useState(statusStates[3])

  const getStatus = useCallback(async (): Promise<IAudit> => {
    const audit: IAudit = await api('/audit/status', { electionId })
    return audit
  }, [electionId])

  const updateAudit = useCallback(async () => {
    setIsLoading(true)
    const audit = await getStatus()
    setAudit(audit)
    setIsLoading(false)
  }, [getStatus])

  useEffect(() => {
    updateAudit()
  }, [updateAudit])

  const round = audit.rounds[audit.rounds.length - 1]

  const [ballots, setBallots] = useState<IBallot[]>(dummyBallots.ballots)

  const getBallots = useCallback(async (): Promise<IBallot[]> => {
    if (audit.jurisdictions.length) {
      const allBallots: IBallot[] = []
      audit.jurisdictions[0].auditBoards!.forEach(async board => {
        const { ballots } = await api(
          `/jurisdiction/${audit.jurisdictions[0].id}/audit-board/${board.id}/round/${round.id}/ballot-list`,
          { electionId }
        )
        allBallots.push(...ballots)
      })
      // return allBallots.length ? allBallots : dummyBallots.ballots
      return allBallots
    } else {
      return []
    }
  }, [electionId, audit.jurisdictions, round])

  const updateBallots = useCallback(async () => {
    setIsLoading(true)
    const ballots = await getBallots()
    setBallots(ballots)
    setIsLoading(false)
  }, [getBallots])

  useEffect(() => {
    updateBallots()
  }, [updateBallots, audit.jurisdictions.length])

  const board:
    | IAuditBoard
    | undefined = audit.jurisdictions[0].auditBoards.find(
    (v: IAuditBoard) => v.id === token
  )

  const nextBallot = (r: string, batchId: string, ballot: string) => () => {
    const ballotIx = ballots.findIndex(
      (b: IBallot) => b.batch.id === batchId && b.position === ballot
    )
    if (ballotIx > -1 && ballots[ballotIx + 1]) {
      const b = ballots[ballotIx + 1]
      history.push(`${url}/round/${r}/batch/${b.batch.id}/ballot/${b.position}`)
    } else {
      history.push(url)
    }
  }

  const previousBallot = (r: string, batchId: string, ballot: string) => () => {
    const ballotIx = ballots.findIndex(
      (b: IBallot) => b.batch.id === batchId && b.position === ballot
    )
    if (ballotIx > -1 && ballots[ballotIx - 1]) {
      const b = ballots[ballotIx - 1]
      history.push(`${url}/round/${r}/batch/${b.batch.id}/ballot/${b.position}`)
    } else {
      history.push(url)
    }
  }

  /* istanbul ignore if */
  if (!board) {
    return (
      <Wrapper>
        <Spinner />
      </Wrapper>
    )
  } else if (board.members.length) {
    return (
      <Wrapper>
        <Switch>
          <Route
            exact
            path="/election/:electionId/board/:token"
            render={({ match: { url } }) => (
              <BoardTable
                isLoading={isLoading}
                setIsLoading={setIsLoading}
                boardName={board.name}
                ballots={ballots}
                round={audit.rounds.length}
                url={url}
              />
            )}
          />
          <Route
            path={url + '/round/:roundId/batch/:batchId/ballot/:ballotId'}
            render={({
              match: {
                params: { roundId, batchId, ballotId },
              },
            }) => (
              <Ballot
                home={url}
                previousBallot={previousBallot(roundId, batchId, ballotId)}
                nextBallot={nextBallot(roundId, batchId, ballotId)}
                contest={audit.contests[0].name}
                roundId={roundId}
                batchId={batchId}
                ballotId={ballotId}
                ballots={ballots}
                boardName={board.name}
              />
            )}
          />
        </Switch>
      </Wrapper>
    )
  } else {
    return (
      <Wrapper>
        <MemberForm
          updateAudit={updateAudit}
          boardName={board.name}
          boardId={board.id}
          jurisdictionName={audit.jurisdictions[0].name}
          jurisdictionId={audit.jurisdictions[0].id}
          electionId={electionId}
        />
      </Wrapper>
    )
  }
}

export default AuditFlow
