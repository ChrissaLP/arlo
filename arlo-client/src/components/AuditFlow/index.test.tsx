import React from 'react'
import { render, wait } from '@testing-library/react'
import { StaticRouter } from 'react-router-dom'
import { routerTestProps } from '../testUtilities'
import AuditFlow from './index'
import { dummyBoard } from './_mocks'
import statusStates from '../AuditForms/_mocks'
import api from '../utilities'

const dummy = (i: number) => {
  statusStates[3].jurisdictions[0].auditBoards = [dummyBoard[i]]
  return statusStates[3]
}

const apiMock = api as jest.Mock<ReturnType<typeof api>, Parameters<typeof api>>

jest.mock('../utilities')

apiMock.mockImplementation(async () => dummy(0))

afterEach(() => {
  apiMock.mockClear()
})

const routeProps = routerTestProps('/election/:electionId/board/:token', {
  electionId: '1',
  token: '123',
})

describe('AuditFlow', () => {
  it('renders correctly', () => {
    const { container } = render(
      <StaticRouter {...routeProps}>
        <AuditFlow {...routeProps} />
      </StaticRouter>
    )
    expect(container).toMatchSnapshot()
  })

  it('fetches initial state from api', async () => {
    const { container } = render(
      <StaticRouter {...routeProps}>
        <AuditFlow {...routeProps} />
      </StaticRouter>
    )
    await wait(() => {
      expect(apiMock).toBeCalled()
      expect(container).toMatchSnapshot()
    })
  })

  it('renders member form', async () => {
    apiMock.mockImplementationOnce(async () => dummy(0))
    const { container, getByText } = render(
      <StaticRouter {...routeProps}>
        <AuditFlow
          {...routeProps}
          dummyID={0}
          testName="member form: dummy 0"
        />
      </StaticRouter>
    )
    await wait(() => {
      expect(apiMock).toBeCalled()
      expect(
        getByText('Member Sign in for Audit Board: Audit Board #1')
      ).toBeTruthy()
      expect(container).toMatchSnapshot()
    })
  })

  it('renders board table with no ballots', async () => {
    apiMock.mockImplementationOnce(async () => dummy(1))
    const { queryByText, getByText } = render(
      <StaticRouter {...routeProps}>
        <AuditFlow
          {...routeProps}
          dummyID={1}
          testName="blank board table: dummy 1"
        />
      </StaticRouter>
    )
    await wait(() => {
      expect(apiMock).toBeCalled()
      expect(getByText('Audit Board #1: Ballot Cards to Audit')).toBeTruthy()
      expect(queryByText('Start Auditing')).toBeFalsy()
    })
  })

  it('renders board table with ballots', async () => {
    apiMock.mockImplementationOnce(async () => dummy(2))
    const { container, getByText } = render(
      <StaticRouter {...routeProps}>
        <AuditFlow
          {...routeProps}
          dummyID={2}
          testName="board table: dummy 1"
        />
      </StaticRouter>
    )
    await wait(() => {
      expect(apiMock).toBeCalled()
      expect(getByText('Audit Board #1: Ballot Cards to Audit')).toBeTruthy()
      expect(getByText('Start Auditing')).toBeTruthy()
      expect(container).toMatchSnapshot()
    })
  })
})
