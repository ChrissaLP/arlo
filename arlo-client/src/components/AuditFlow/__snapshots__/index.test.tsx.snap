// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuditFlow renders correctly 1`] = `
<div>
  <div
    class="sc-eNQAEJ jDJHSN"
  />
</div>
`;
